test/versioned/kafkajs/kafka.tap.js

```diff
@@ -43,24 +47,172 @@ tap.afterEach(async (t) => {
   await t.context.producer.disconnect()
 })
 
-tap.test('stub', async (t) => {
-  const { consumer, producer, topic } = t.context
+tap.test('send records correctly', (t) => {
+  t.plan(4)
+
+  const { agent, consumer, producer, topic } = t.context
   const message = 'test message'
 
-  await consumer.subscribe({ topics: [topic], fromBeginning: true })
-  const testPromise = new Promise((resolve) => {
-    consumer.run({
-      eachMessage: async ({ message: actualMessage }) => {
-        t.equal(actualMessage.value.toString(), message)
-        resolve()
-      }
+  agent.on('transactionFinished', (tx) => {
+    const name = `MessageBroker/Kafka/Topic/Produce/Named/${topic}`
+    const segment = tx.agent.tracer.getSegment()
+
+    const foundSegment = segment.children.find((s) => s.name.endsWith(topic))
+    t.equal(foundSegment.name, name)
+
+    const metric = tx.metrics.getMetric(name)
+    t.equal(metric.callCount, 1)
+
+    t.end()
+  })
+
+  helper.runInTransaction(agent, async (tx) => {
+    await consumer.subscribe({ topic, fromBeginning: true })
+    const promise = new Promise((resolve) => {
+      consumer.run({
+        eachMessage: async ({ message: actualMessage }) => {
+          t.equal(actualMessage.value.toString(), message)
+          t.match(actualMessage.headers['x-foo'].toString(), 'foo')
+          resolve()
+        }
+      })
+    })
+    await utils.waitForConsumersToJoinGroup({ consumer })
+    await producer.send({
+      acks: 1,
+      topic,
+      messages: [
+        {
+          key: 'key',
+          value: message,
+          headers: {
+            'x-foo': 'foo'
+          }
+        }
+      ]
+    })
+    await promise
+
+    tx.end()
+  })
+})
+
+tap.test('send passes along DT headers', (t) => {
+  // The intent of this test is to verify the scenario:
+  //
+  // 1. A service receives a request
+  // 2. The service builds a payload for Kafka
+  // 3. The produced Kafka data includes the distributed trace data that was
+  // provided to the service handling the request.
+
+  t.plan(5)
+
+  const now = Date.now
+  Date.now = () => 1717426365982
+  t.teardown(() => {
+    Date.now = now
+  })
+
+  const { agent, consumer, producer, topic } = t.context
+  const messages = ['one', 'two', 'three']
+
+  // These agent.config lines are utilized to simulate the inbound
+  // distributed trace that we are trying to validate.
+  agent.config.account_id = 'account_1'
+  agent.config.primary_application_id = 'app_1'
+  agent.config.trusted_account_key = 42
+
+  agent.on('transactionFinished', (tx) => {
+    t.equal(tx.isDistributedTrace, true)
+
+    const headers = {}
+    tx.traceContext.addTraceContextHeaders(headers)
```

Comment:
We can merge this but I'll clean up when consumer instrumentation is added

Reply:
What needs to be cleaned up? The method is poorly named. It is not adding headers to the transaction. It is retrieving them from the transaction and adding them to the passed in object.